### ISSUE TITLE
building jaxlib 0.1.71 from source with GCC installed in non-standard location fails

### ISSUE THREAD

**Issue by boegel** [google/jax#7842]
When building `jaxlib` 0.1.71 from source with a GCC 10.3 installation that is located in a non-standard location (say `/apps/GCC/10.3`), I'm running into the following build failure:

```
ERROR: /tmp/tmpeyx0odma-bazel/bc7c284e89e194c4746b754cd3310461/external/llvm-project/mlir/BUILD.bazel:144:18: TdGenerate external/llvm-project/mlir/include/mlir/IR/BuiltinAttributes.h.inc failed (Exit 1): mlir-tblgen failed: error executing command
  (cd /tmp/tmpeyx0odma-bazel/bc7c284e89e194c4746b754cd3310461/execroot/__main__ && \
  exec env - \
  bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen --gen-attrdef-decls external/llvm-project/mlir/include/mlir/IR/BuiltinAttributes.td -I external/llvm-project/mlir/include -I bazel-out/k8-opt/bin/external/llvm-project/mlir/include -I external/llvm-project/ -I bazel-out/k8-opt/bin/external/llvm-project/ -I external/llvm-project/mlir/include/mlir/IR -I bazel-out/k8-opt/bin/external/llvm-project/mlir/include/mlir/IR -o bazel-out/k8-opt/bin/external/llvm-project/mlir/include/mlir/IR/BuiltinAttributes.h.inc)
Execution platform: @local_execution_config_platform//:platform
bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen)
bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen)
bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen)
bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen: /lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/mlir/mlir-tblgen)
Target //build:build_wheel failed to build
ERROR: /tmp/tmpeyx0odma-bazel/bc7c284e89e194c4746b754cd3310461/external/org_tensorflow/tensorflow/compiler/xla/python/BUILD:518:17 TdGenerate external/llvm-project/mlir/include/mlir/IR/BuiltinAttributes.h.inc failed (Exit 1): mlir-tblgen failed: error executing command
```

We have seen this type of problem before: it occurs when the environment in which a tool that was built earlier (`mlir-tblgen` in this case) is reset by Bazel when using that tool: because `$LD_LIBRARY_PATH` no longer contains the location to the correct `libstdc++.so.6` the one from the OS (`/lib64/libstdc++.so.6`) is picked up, but that's an older version than the one used to build `mlir-tblgen`, and hence it fails to run.

We also ran into this problem when building `jaxlib` 0.1.70 from source, and then we came up with a [patch](https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/j/jax/TensorFlow-4039feeb743bc42cd0a3d8146ce63fc05d23eb8d_mlir-build-env.patch) that enables `use_default_shell_env`, which effectively avoids the environment reset.

For `jaxlib` 0.1.71 that fix doesn't work anymore however, because the `third_party/mlir/tblgen.bzl` file is no longer there in the [TensorFlow commit](https://github.com/tensorflow/tensorflow/commits/b5b1ff47ad250c3e38dcadef5f6bc414b0a533ee) being used, and I'm not sure where a similar fix can be applied to avoid the environment reset.

I know that @Flamefire has been fighting issues like this several times when building TensorFlow from source, so perhaps he knows how to fix this issue...

**Comment by yashk2810** [comment]
When I build jaxlib with RBE where the compiler location is present somewhere else, there are some env_vars that are specified in the configs to help bazel find the right location. Maybe you can try them out?

https://github.com/tensorflow/toolchains/blob/master/toolchains/remote_config/rbe_config.bzl#L19-L31 (the `compiler` parameter is how the path is specified)

I can see env_vars like `BAZEL_COMPILER`, etc.

I don't know if it will work but worth a shot I guess.

You can also set LD_LIBRARY_PATH by passing it as an action_env to the bazel command.

`--action_env=LD_LIBRARY_PATH="/usr/local/cuda:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/tensorrt/lib"` for example

**Comment by Flamefire** [comment]
> You can also set LD_LIBRARY_PATH by passing it as an action_env to the bazel command.

That doesn't work at all when `use_default_shell_env = False` is not also used in the run-rule, which in this case is in the llvm repo. As you can see in the above output *no* environment variables are passed to the command in the current state.   
Previously the rule was in the TensorFlow repo where we patched it with https://github.com/easybuilders/easybuild-easyconfigs/blob/b2eb591fb6ccdf5bff42beef84797e6d40fca95b/easybuild/easyconfigs/j/jax/TensorFlow-4039feeb743bc42cd0a3d8146ce63fc05d23eb8d_mlir-build-env.patch

But now it needs a patch of the llvm repo which is downloaded by TensorFlow which is downloaded by jax. Great!
I opened a merge request in the llvm repo: https://reviews.llvm.org/D109873
TLDR: Bazel and environments is a mess

**Comment by yashk2810** [comment]
I am closing this since this is a very old issue. JAX and jaxlib are at 0.4.20. Can you try at HEAD now?

Feel free to reopen or open a new issue if this error keeps happening.
